repose-aggregator/components/filters/ip-classification/src/main/scala/org/openrepose/filters/ipclassification/IPClassificationFilter.scala

```diff
@@ -0,0 +1,142 @@
+  private var groupHeaderName: String = _
+  private var groupHeaderQuality: Double = _
+  private var userHeaderName: String = _
+  private var userHeaderQuality: Double = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    configName = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+    logger.info(s"Initializing using config $configName")
+    val xsdURL: URL = getClass.getResource("/META-INF/config/schema/ip-classification.xsd")
+
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      configName,
+      xsdURL,
+      this,
+      classOf[IpClassificationConfig]
+    )
+  }
+
+  override def destroy(): Unit = configurationService.unsubscribeFrom(configName, this.asInstanceOf[UpdateListener[_]])
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    if (!initialized) {
+      logger.error("IP Classification filter has not yet initialized...")
+      servletResponse.asInstanceOf[HttpServletResponse].sendError(500)
+    } else {
+      logger.trace("IP Classification filter handling request...")
+      val request = MutableHttpServletRequest.wrap(servletRequest.asInstanceOf[HttpServletRequest])
+
+      getClassificationLabel(servletRequest.getRemoteAddr).foreach { label =>
+        request.addHeader(groupHeaderName, s"$label;q=$groupHeaderQuality")
+      }
+
+      //Always set the user header name to the current IP address
+      request.addHeader(userHeaderName, s"${servletRequest.getRemoteAddr};q=$userHeaderQuality")
+
+      logger.trace("IP Classification filter passing request...")
+      filterChain.doFilter(request, servletResponse)
+
+      logger.trace("IP Classification filter returning response...")
+    }
+  }
+
+  def getClassificationLabel(ipAddress: String): Option[String] = {
+    cidrList.get().find { labeledCidr =>
+      labeledCidr.cidr.isInRange(ipAddress)
+    } map { foundLabeledCidr =>
+      foundLabeledCidr.label
+    }
+  }
+
+  override def configurationUpdated(classificationConfig: IpClassificationConfig): Unit = {
+    //Update all the CIDRs
+    //Create a new list to replace the old one
+
+    import scala.collection.JavaConversions._
+    val groups = classificationConfig.getGroup.toList
+
+    /**
+     * This guy builds a List[List[LabeledCIDR]] I flat map it to remove that extra list, so it's just a
+     * List[LabeledCIDR]. I suppose I could separately transform the classification lines into lists, and then combine
+     * them, but this does the same thing
+     */
+    val replacementCidrList: List[LabeledCIDR] = groups.flatMap { group =>
+      val label = group.getName
+      group.getCidrIp.map { cidr =>
+        LabeledCIDR(label, new CIDRUtils(cidr))
+      }
+    }
+
+    //Replace our object! Atomically
+    cidrList.set(replacementCidrList)
+
+    //Blergh, no useful defaults in XSD when I add complex types :(
```

Comment:
If the element is not present, then it doesn't provide defaults; it returns `null`.